src/foam/util/test/UIDUniquenessTest.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.util.test',
+  name: 'UIDUniquenessTest',
+  extends: 'foam.nanos.test.Test',
+
+  javaImports: [
+    'foam.core.ContextAware',
+    'foam.core.X',
+    'foam.util.UIDGenerator',
+    'java.io.IOException',
+    'java.util.Collections',
+    'java.util.concurrent.CountDownLatch',
+    'java.util.HashSet'
+  ],
+
+  properties: [
+    {
+      class: 'Int',
+      name: 'size',
+      value: 100000
+    },
+    {
+      name: 'isDuplicate',
+      class: 'Boolean',
+      value: false
+    },
+    {
+      name: 'isNotDuplicate',
+      class: 'Boolean',
+      value: false
+    }
+  ],
+
+  methods: [
+    {
+      name: 'runTest',
+      javaCode: `
+        /*
+          Test case 1:
+          When machine Id is same(isDifferent = false),
+          the test passes if a duplicate is found.
+        */
+        UIDUniquenessTest_UIDDuplicateFoundTest(x, false);
+
+        /*
+          Test case 2:
+          When machine Id is different(isDifferent = true),
+          the test passes if no duplicate is found.
+        */
+        UIDUniquenessTest_UIDDuplicateNotFoundTest(x, true);
+      `
+    },
+    {
+      name: 'addUids',
+      args: 'Context x, boolean isDifferent, boolean isDuplicate, boolean isNotDuplicate',
+      documentation: 'Create two threads and add uids to each set',
+      javaCode: `
+        // To store unique ids into a set in the 1st thread.
+        var uids = Collections.synchronizedSet(new HashSet());
+        var numOfThread = 2;
+        final CountDownLatch latch = new CountDownLatch(numOfThread);
+
+        for ( int i = 0; i < numOfThread; i++ ) {
+          final int tno = i;
+          Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+              var uidGenerator = new UIDGenerator.Builder(x).setSalt("foobar").build();
+              if ( isDifferent ) {
+                uidGenerator.setMachineId(uidGenerator.getMachineId() + tno);
+              }
+              try {
+                for ( int j = 0; j < getSize(); j++ ) {
+                  String uid = uidGenerator.getNextString();
+                  // Add a unique id into a set according to tno.
+                  if ( tno == 0 ) {
+                    uids.add(uid);
```

Comment:
There is no guarantee that thread 0 will complete before thread 1. 
If you want to go with one Collection, then use a Map<uid, count>.
ConcurrentHashMap, and count would an AtomicLong. 
Then after the runs, walk the map and check for count of 2, which indicates duplicate. 
But with one Collection, you can't tell if the two runs generated the same number of ids.  If a uidgenerator itself generated duplicates then it would show as a count of 2. 
So I think best to have two maps. 
Then
- check for same number of ids.
- walk one map and check contains in the other - for duplicates,  Also during this, you can check the count for greater than 1.  

summary
to check for same number generated - two collections
to check for duplicates - can use one or two collections.